File: app/src/main/java/com/mgaetan89/showsrage/fragment/SettingsServerApiKeyFragment.java
package com.mgaetan89.showsrage.fragment;

import android.content.SharedPreferences;
import android.preference.Preference;
import android.preference.PreferenceManager;
import android.preference.PreferenceScreen;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;

import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.model.ApiKey;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.lang.ref.WeakReference;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class SettingsServerApiKeyFragment extends SettingsServerFragment {
	@Override
	public boolean onPreferenceTreeClick(PreferenceScreen preferenceScreen, @NonNull Preference preference) {
		if ("get_api_key_action".equals(preference.getKey())) {
			this.getApiKey();

			return true;
		}

		return super.onPreferenceTreeClick(preferenceScreen, preference);
	}

	@Override
	protected int getTitleResourceId() {
		return R.string.api_key;
	}

	@Override
	protected int getXmlResourceFile() {
		return R.xml.settings_server_api_key;
	}

	private void getApiKey() {
		SickRageApi.Companion.getInstance().init(PreferenceManager.getDefaultSharedPreferences(this.getActivity()));

		String username = this.getPreferenceValue("server_username", null);
		String password = this.getPreferenceValue("server_password", null);

		SickRageApi.Companion.getInstance().getServices().getApiKey(username, password, new ApiKeyCallback(this));
	}

	private static final class ApiKeyCallback implements Callback<ApiKey> {
		private WeakReference<SettingsFragment> fragmentReference = null;

		private ApiKeyCallback(SettingsFragment fragment) {
			this.fragmentReference = new WeakReference<>(fragment);
		}

		@Override
		public void failure(RetrofitError error) {
			this.showApiKeyResult(null);
		}

		@Override
		public void success(ApiKey apiKey, Response response) {
			if (apiKey.isSuccess()) {
				this.showApiKeyResult(apiKey.getApiKey());
			} else {
				this.showApiKeyResult(null);
			}
		}

		private void setPreferenceValue(SettingsFragment fragment, String key, String value) {
			SharedPreferences sharedPreferences = fragment.getPreferenceManager().getSharedPreferences();
			sharedPreferences.edit().putString(key, value).apply();
		}

		private void showApiKeyResult(@Nullable String apiKey) {
			SettingsFragment fragment = this.fragmentReference.get();

			if (fragment == null) {
				return;
			}

			int messageId;

			if (TextUtils.isEmpty(apiKey)) {
				messageId = R.string.get_api_key_error;
			} else {
				this.setPreferenceValue(fragment, "api_key", apiKey);
				fragment.findPreference("api_key").setSummary(apiKey);

				messageId = R.string.get_api_key_success;
			}

			new AlertDialog.Builder(fragment.getActivity())
					.setCancelable(true)
					.setMessage(messageId)
					.setPositiveButton(android.R.string.ok, null)
					.show();
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/SettingsServerApiKeyFragment.kt
package com.mgaetan89.showsrage.fragment

import android.support.v7.app.AlertDialog
import android.support.v7.preference.Preference
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.extension.Fields
import com.mgaetan89.showsrage.extension.getPreferences
import com.mgaetan89.showsrage.extension.getServerPassword
import com.mgaetan89.showsrage.extension.getServerUsername
import com.mgaetan89.showsrage.model.ApiKey
import com.mgaetan89.showsrage.network.SickRageApi
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response
import java.lang.ref.WeakReference

class SettingsServerApiKeyFragment : SettingsServerFragment() {
	override fun onPreferenceTreeClick(preference: Preference?): Boolean {
		if ("get_api_key_action" == preference?.key) {
			this.getApiKey()

			return true
		}

		return super.onPreferenceTreeClick(preference)
	}

	override fun getTitleResourceId() = R.string.api_key

	override fun getXmlResourceFile() = R.xml.settings_server_api_key

	private fun getApiKey() {
		val preferences = this.activity.getPreferences()

		SickRageApi.instance.init(preferences)

		val username = preferences.getServerUsername()
		val password = preferences.getServerPassword()

		SickRageApi.instance.services?.getApiKey(username, password, ApiKeyCallback(this))
	}

	private class ApiKeyCallback(fragment: SettingsFragment) : Callback<ApiKey> {
		private val fragmentReference = WeakReference(fragment)

		override fun failure(error: RetrofitError?) {
			this.showApiKeyResult(null)
		}

		override fun success(apiKey: ApiKey?, response: Response?) {
			if (apiKey?.success ?: false) {
				this.showApiKeyResult(apiKey?.apiKey)
			} else {
				this.showApiKeyResult(null)
			}
		}

		private fun setPreferenceValue(fragment: SettingsFragment, key: String, value: String) {
			val sharedPreferences = fragment.preferenceManager.sharedPreferences
			sharedPreferences.edit().putString(key, value).apply()
		}

		private fun showApiKeyResult(apiKey: String?) {
			val fragment = this.fragmentReference.get() ?: return
			val activity = fragment.activity ?: return
			val messageId: Int

			if (apiKey.isNullOrEmpty()) {
				messageId = R.string.get_api_key_error
			} else {
				this.setPreferenceValue(fragment, Fields.API_KEY.field, apiKey!!)
				fragment.findPreference(Fields.API_KEY.field).summary = apiKey

				messageId = R.string.get_api_key_success
			}

			AlertDialog.Builder(activity)
					.setCancelable(true)
					.setMessage(messageId)
					.setPositiveButton(android.R.string.ok, null)
					.show()
		}
	}
}
